File: src/ros2_bridge/mappers/sensor_msgs/point_field.rs
//! Typed mapper for `sensor_msgs/msg/PointField`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn point_field_to_bus(
    msg: ros_env::sensor_msgs::msg::PointField,
) -> crate::sensor_msgs::msg::v1::PointField {
    crate::sensor_msgs::msg::v1::PointField {
        name: crate::ros2_bridge::mappers::convert::from_ros_string(msg.name),
        offset: msg.offset.into(),
        datatype: u32::from(msg.datatype),
        count: msg.count.into(),
    }
}

pub(crate) fn point_field_to_ros(
    bus: crate::sensor_msgs::msg::v1::PointField,
) -> ros_env::sensor_msgs::msg::PointField {
    ros_env::sensor_msgs::msg::PointField {
        name: crate::ros2_bridge::mappers::convert::to_ros_string(bus.name),
        offset: bus.offset as _,
        datatype: bus.datatype as u8,
        count: bus.count as _,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsPointFieldMapper;

impl TypedTopicMapper for SensorMsgsPointFieldMapper {
    type Ros = ros_env::sensor_msgs::msg::PointField;
    type Bus = crate::sensor_msgs::msg::v1::PointField;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(point_field_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(point_field_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/point_field.py
"""Generated mapper for `sensor_msgs/msg/PointField`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.sensor_msgs.msg.v1 import PointField as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def point_field_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.name = str(msg.name)
    bus.offset = msg.offset
    bus.datatype = int(msg.datatype)
    bus.count = msg.count
    return bus


def point_field_to_ros(bus):
    from sensor_msgs.msg import PointField as RosMsg

    out = RosMsg()
    out.name = str(bus.name)
    out.offset = bus.offset
    out.datatype = int(bus.datatype)
    out.count = bus.count
    return out


class SensorMsgsPointFieldMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from sensor_msgs.msg import PointField as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return point_field_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return point_field_to_ros(bus)
